src/core/lib/surface/call.cc

```diff
@@ -62,170 +64,277 @@
 #include "src/core/lib/transport/error_utils.h"
 #include "src/core/lib/transport/transport.h"
 
-/** The maximum number of concurrent batches possible.
-    Based upon the maximum number of individually queueable ops in the batch
-    api:
-      - initial metadata send
-      - message send
-      - status/close send (depending on client/server)
-      - initial metadata recv
-      - message recv
-      - status/close recv (depending on client/server) */
-#define MAX_CONCURRENT_BATCHES 6
-
-struct batch_control {
-  batch_control() = default;
-
-  grpc_call* call = nullptr;
-  grpc_transport_stream_op_batch op;
-  /* Share memory for cq_completion and notify_tag as they are never needed
-     simultaneously. Each byte used in this data structure count as six bytes
-     per call, so any savings we can make are worthwhile,
-
-     We use notify_tag to determine whether or not to send notification to the
-     completion queue. Once we've made that determination, we can reuse the
-     memory for cq_completion. */
-  union {
-    grpc_cq_completion cq_completion;
-    struct {
-      /* Any given op indicates completion by either (a) calling a closure or
-         (b) sending a notification on the call's completion queue.  If
-         \a is_closure is true, \a tag indicates a closure to be invoked;
-         otherwise, \a tag indicates the tag to be used in the notification to
-         be sent to the completion queue. */
-      void* tag;
-      bool is_closure;
-    } notify_tag;
-  } completion_data;
-  grpc_closure start_batch;
-  grpc_closure finish_batch;
-  std::atomic<intptr_t> steps_to_complete{0};
-  AtomicError batch_error;
-  void set_num_steps_to_complete(uintptr_t steps) {
-    steps_to_complete.store(steps, std::memory_order_release);
+grpc_core::TraceFlag grpc_call_error_trace(false, "call_error");
+grpc_core::TraceFlag grpc_compression_trace(false, "compression");
+
+namespace grpc_core {
+
+class Call : public CppImplOf<Call, grpc_call> {
```

Comment:
quick question: What's the benefit of CppImplOf?

Reply:
So when we have C types that we want to implement in C++ we get some design choices...

Choice 1:
```
struct grpc_call {
  grpc_call(...);
  void foo();
  
 private:
  int member1_;
};
void grpc_call_foo(grpc_call* c) { c->foo(); }
```
Here we land the C++ implementation directly in the C struct... we end up with something that performs nicely, but the implementation is a confusing mixture of styles, and the resulting code looks like someone vomited a great deal of `grpc_core::`'s all over the place.

Choice 2:
```
namespace grpc_core {
class Call {
 public:
  void foo();
 private:
  grpc_call* outer_;
};
struct grpc_call {
  unique_ptr<grpc_core::Call> inner;
};
void grpc_call_foo(grpc_call* c) { c->inner->foo(); }
```
Here we wrap the inner C++ type with a C struct, and use a member variable to get back out again. It's fiddly to construct, but more importantly costs an extra allocation and two additional pointers for each wrapped type - and necessitates touching two cachelines for every access to the object from outside. We've made the code more idiomatic and typesafe - great for us - but we've potentially sacrificed some performance for users.

Choice 3:
```
struct grpc_call;
namespace grpc_core {
class Call { ... };
}
void grpc_call_foo(grpc_call* c) { reinterpret_cast<grpc_core::Call*>(c)->foo(); }
```
Here we say that grpc_call is just a tag, and we reinterpet_cast back and forth everywhere we might need.
It works... and most of the code ends up looking nice and idiomatic, except at our boundary points - reasonable people avoid writing reinterpret_cast frequently - one might cast to the wrong type with disastrous consequences, it's just ugly, and let's not do that.
This is on par with the best possible performance though, so we're getting closer!

And so finally, Choice 4:
```
struct grpc_call;
namespace grpc_core {
class Call : public CppImplOf<Call, grpc_call> { ... };
}
void grpc_call_foo(grpc_call* c) { grpc_core::Call::FromC(c)->foo(); }
```
We've essentially taken Choice 3, and buried the reinterpret_casts inside CppImplOf.
In doing so, we've limited the number of ways we can mess things up with reinterpret_cast (it's quite hard to misuse), and our code looks pretty dang idiomatic - AND we don't have the dual allocation/cache pollution problems of Choice 2.